### DISCUSSION TITLE
storage

### DISCUSSION THREAD

**Discussion by ouyangqr** [mastodon/mastodon#27935] [General]
How to connect file storage with the file server built by myself, and the file server built by fastdfs I used

**Comment by Gargron** [comment]
Mastodon supports:

- local file storage
- S3 API
- Swift API

If fastdfs works as a mounted volume on the system, then you can simply use `PAPERCLIP_ROOT_PATH` to make Mastodon store uploaded files on that volume (and you'll need to create a `location` in nginx that reads from that volume). Config options:

https://docs.joinmastodon.org/admin/config/#cdn

**Comment by unitof** [comment]
What `location` match will guarantee I only use that volume for media files stored by Paperclip?

I tried adding a `root` directive to the `location ~ ^/(test|emoji|packs|system/accounts/avatars|system/media_attachments/files)` block in nginx site conf, but that means 404s for default Mastodon assets & logos in the `public` dir.
